keps/0029-operator-dependencies.md

```diff
@@ -0,0 +1,282 @@
+
+### Goals
+
+Dependencies can be a complex topic. This KEP is not trying to boil the dependency ocean but rather limits itself to installation dependencies only, i.e. a set of `Operators/Instances` being installed together and removed together as a unit.
+
+### Non-Goals
+
+Dependency on an already running `Instance` is a non-goal. It is easy to imagine a situation when a new operator (e.g Kafka) may want to depend on the existing Zookeeper instance. However, such life-cycle dependency presents major challenges e.g. what happens when Zookeeper is removed? What happens when Zookeeper is upgraded, and the new version is incompatible with the current Kafka `Instance`? How can we ensure the compatibility? This KEP deliberately ignores this area and instead focuses on installation dependencies. Additionally, this KEP does not address output variables or referencing `Instance` resources.
+
+## Proposal
+
+KUDO operators **already have** a mechanism to deal with installation dependencies called [plans, phases, and steps](https://kudo.dev/docs/developing-operators/plans.html#overview) with serial or parallel execution strategy. This mechanism is already powerful enough to express any dependency hierarchy including transitive dependencies (see more about this in [implementation details](#implementation-details)). The core of this proposal is to reuse this mechanism and extend it with the ability to install operators.
+
+### Implementation details
+
+#### Operator Task
+
+A new task kind `Operator` is introduced which extends `operator.yaml` with the ability to install dependencies. Let's take a look at the Kafka+Zookeeper example:
+
+```yaml
+apiVersion: kudo.dev/v1beta1
+name: "kafka"
+operatorVersion: "1.3.1"
+kudoVersion: 0.12.0
+kubernetesVersion: 1.14.8
+appVersion: 2.5.0
+url: https://kafka.apache.org/
+tasks:
+  - name: zookeeper-operator
+    kind: Operator
+    spec:
+        package: zookeeper
+
+  ...
+
+plans:
+  deploy:
+    strategy: serial
+    phases:
+      - name: deploy-zookeeper
+        strategy: serial
+        steps:
+          - name: zookeeper
+            tasks:
+              - zookeeper-operator
+      - name: deploy-kafka
+        strategy: serial
+        steps:
+          ...
+```
+
+The `zookeeper-operator` task specification is equivalent to `kudo install zookeeper` CLI command which installs the Zookeeper package from the official repo. Here is a complete `Operator` task specification:
+
+```yaml
+tasks:
+- name: demo
+  kind: Operator
+  spec:
+    package: # required, either repo package name, local package folder or an URL to package tarball
+    repo: # optional, name of local repository configuration to use
+    appVersion: # optional, a specific app version in the official repo, defaults to the most recent one
+    operatorVersion: # optional, a specific operator version in the official repo, defaults to the most recent one
+    instanceName: # optional, the instance name
```

Comment:
Good point, It might be valuable to allow prefixes or some other linkage to the group. If I name my top-level operator `abc`, the dependency might/should have the name `abc-zookeeper`

Reply:
But that's at the hand of the operator developer, right? If you want, you can name your instances accordingly. If not, KUDO will take the default name, like it does today.

Reply:
Then we either have to extend the defaulting (i.e. if the instance name `zookeeper-instance` is already taken, use `zookeeper-instance1` or provide a way for the operator user to configure this.

Otherwise we run into the problem of having multiple operators that use the same dependency:

Operator AA uses zk and doesn't define an instanceName. zk gets installed as `zookeeper-instance`.
Operator BB uses zk and doesn't define an instanceName. zk wants to install as `zookeeper-instance` but it's already there and it fails. The user now has no way to install Operator BB. Or am I missing something?

Making the instanceName required may help, but... Not sure if that's the best way.

Reply:
Additionally - If i want to install AA twice, with different instance names, I can do that for the top-level operator. But if the dependencies don't rely on the top-level-instance name, they will always clash.

Reply:
These are all good points. Wdyt about making dependency instance names hierarchical? E.g. if the top-level operator instance is named `kafka-prod` then it's ZK instance is named `kakfa-prod.zk-instance`? This way, it is also immediately clear in the `$ k get instances` where the instance belongs to?

Reply:
Yeah, I'd say that'd be a good idea. Not sure which separator we should use. `.` vs `-` vs `_`, but some kind of hierarchy would be good.

Reply:
Would be nice, but I'm not sure the schema for `metadata.name` will like the dots/underscores or the resulting long names in case of deeply nested operators...

Reply:
From the ̀[docs](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-label-names) an object name should:

- contain no more than 253 characters
- contain only lowercase alphanumeric characters, ‘-’ or ‘.’
- start with an alphanumeric character
- end with an alphanumeric character

Dots are fine but 253 might become an issue somewhere down the road. We could encourage short names for dependencies and drop the `-instance` suffix.